tests/std/tests/P0896R4_ranges_alg_generate/test.cpp

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+
+#include <range_algorithm_support.hpp>
+
+constexpr void smoke_test() {
+    using ranges::generate, ranges::iterator_t;
+    using std::same_as;
+
+    {
+        int output[]    = {13, 42, 1367};
+        const int value = 7;
+        auto result     = generate(std::begin(output), std::end(output), []() { return value; });
+        for (auto elem : output) {
+            assert(elem == value);
+        }
+        assert(result == std::end(output));
+    }
+    {
+        int output[]    = {13, 42, 1367};
+        const int value = 13;
+        auto result     = generate(output, []() { return value; });
+        for (auto elem : output) {
+            assert(elem == value);
+        }
+        assert(result == std::end(output));
+    }
+    {
+        int output[]          = {13, 42, 1367};
+        int calls_to_generate = -1;
+        auto result           = generate(output, [&calls_to_generate]() {

```

Comment:
This is great! It made me realize that there's one more way that the implementation could be incorrect, and how to verify it cheaply. It's tempting for us to mark things as `const`, but we really don't want to mark the generator function object as `const` in the algorithm (which would prohibit function objects with non-`const` function call operators).

Here, the lambda function has a `const` function call operator (like all lambdas receive by default); it modifies `calls_to_generate` through the stored reference. That's fine, except that it wouldn't detect the potential implementation problem of marking the function object as `const`. Since we don't need `calls_to_generate` outside the lambda, we could instead use an init-capture and make the lambda `mutable` (which removes `const` from its function call operator). The behavior would be the same, but this would verify that `generate()` is invoking the function object without adding `const` to it. I believe the syntax for this is:

```suggestion
        auto result           = generate(output, [calls_to_generate = -1]() mutable {
```

(This will also affect clang-format.)